### PR TITLE
PYTHON-3347 Test against Python 3.11 prerelease

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

**Review by ShaneHarvey — APPROVED** [review]

**Comment by ShaneHarvey** [comment]
Assuming the 3.11 tests pass let's add the 3.11 trove classifier too.

**Comment by blink1073** [comment]
Failures that were pre-existing in recent [build](https://spruce.mongodb.com/version/mongo_python_driver_774154e934509dd0c7d854f6491304a295845dd6/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC):

- test_dns_record_lookup_empty (test_srv_polling.TestSrvPolling)
- test_dns_failures (test_srv_polling.TestSrvPolling)
- test_srv_polling.TestSrvPolling.test_dns_failures
- test_srv_polling.TestSrvPolling.test_dns_record_lookup_empty


New failures:

- test_corpus (test.test_encryption.TestCorpus.test_corpus) [src](https://spruce.mongodb.com/task/mongo_python_driver_tests_windows_encryption__platform~windows_64_vsMulti_small_auth_ssl~auth_ssl_python_version_windows~3.11_encryption~encryption_crypt_shared_test_6.0_standalone_patch_774154e934509dd0c7d854f6491304a295845dd6_633f2e05e3c3317f31324fbd_22_10_06_19_35_34/logs?execution=0&sortBy=STATUS&sortDir=ASC)
-   `[2022/10/06 21:02:41.676] Running command 'shell.exec' in "run enterprise auth tests" (step 2 of 2).
 [2022/10/06 21:09:54.828] Command ''shell.exec' in "run enterprise auth tests"' failed: shell script encountered problem: exit code 1.` [src](https://spruce.mongodb.com/task/mongo_python_driver_tests_windows_enterprise_auth__platform~windows_64_vsMulti_small_auth~auth_python_version_windows~3.11_test_enterprise_auth_patch_774154e934509dd0c7d854f6491304a295845dd6_633f2e05e3c3317f31324fbd_22_10_06_19_35_34/logs?execution=0&sortBy=STATUS&sortDir=ASC)

**Comment by ShaneHarvey** [comment]
```
 [2022/10/06 21:40:11.633] ERROR: test_corpus (test.test_encryption.TestCorpus.test_corpus)
 [2022/10/06 21:40:11.633] ----------------------------------------------------------------------
 [2022/10/06 21:40:11.633] Traceback (most recent call last):
 [2022/10/06 21:40:11.633]   File "C:\data\mci\ccc3b9397cdce05218603e98ad57b081\src\pymongo\encryption.py", line 76, in _wrap_encryption_errors
 [2022/10/06 21:40:11.633]     yield
 [2022/10/06 21:40:11.633]   File "C:\data\mci\ccc3b9397cdce05218603e98ad57b081\src\pymongo\encryption.py", line 364, in encrypt
 [2022/10/06 21:40:11.633]     encrypted_cmd = self._auto_encrypter.encrypt(database, encoded_cmd)
 [2022/10/06 21:40:11.633]                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 [2022/10/06 21:40:11.633]   File "Z:\data\mci\ccc3b9397cdce05218603e98ad57b081\src\venv-encryption\Lib\site-packages\pymongocrypt\auto_encrypter.py", line 44, in encrypt
 [2022/10/06 21:40:11.633]     return run_state_machine(ctx, self.callback)
 [2022/10/06 21:40:11.633]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 [2022/10/06 21:40:11.633]   File "Z:\data\mci\ccc3b9397cdce05218603e98ad57b081\src\venv-encryption\Lib\site-packages\pymongocrypt\state_machine.py", line 150, in run_state_machine
 [2022/10/06 21:40:11.633]     callback.kms_request(kms_ctx)
 [2022/10/06 21:40:11.633]   File "C:\data\mci\ccc3b9397cdce05218603e98ad57b081\src\pymongo\encryption.py", line 142, in kms_request
 [2022/10/06 21:40:11.633]     data = conn.recv(kms_context.bytes_needed)
 [2022/10/06 21:40:11.633]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 [2022/10/06 21:40:11.633]   File "C:\python\Python311\Lib\ssl.py", line 1263, in recv
 [2022/10/06 21:40:11.633]     return self.read(buflen)
 [2022/10/06 21:40:11.633]            ^^^^^^^^^^^^^^^^^
 [2022/10/06 21:40:11.633]   File "C:\python\Python311\Lib\ssl.py", line 1136, in read
 [2022/10/06 21:40:11.633]     return self._sslobj.read(len)
 [2022/10/06 21:40:11.633]            ^^^^^^^^^^^^^^^^^^^^^^
 [2022/10/06 21:40:11.633] TimeoutError: The read operation timed out
```

Interesting, is this a fluke or does it happen consistently?

**Comment by blink1073** [comment]
I kicked the task and it [passed](https://spruce.mongodb.com/task/mongo_python_driver_tests_windows_encryption__platform~windows_64_vsMulti_small_auth_ssl~auth_ssl_python_version_windows~3.11_encryption~encryption_crypt_shared_test_6.0_standalone_patch_774154e934509dd0c7d854f6491304a295845dd6_633f2e05e3c3317f31324fbd_22_10_06_19_35_34/logs?execution=1)

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_bson.py [hunk 1]
```diff
@@ -982,6 +982,10 @@ def test_codec_options_repr(self):
             "fallback_encoder=None), "
             "datetime_conversion=DatetimeConversion.DATETIME)"
         )
+        # The repr() of an enum changed in Python 3.11.
+        if sys.version_info >= (3, 11):
+            r = r.replace("DatetimeConversion.DATETIME", str(DatetimeConversion.DATETIME))
```
[the remark this answers is not in the document]
Comment (blink1073) [reply test/test_bson.py:987]: We're actually using the `str()` representation, and it is an intentional change:

[`IntEnum` used to subclass `int, Enum` and now it subclasses `int, ReprEnum`.](https://docs.python.org/3.11/library/enum.html#enum.IntEnum): 

"Changed in version 3.11: __str__() is now int.__str__() to better support the replacement of existing constants use-case. __format__() was already int.__format__() for that same reason."

[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply test/test_bson.py:987]: can we use format (f-string) instead?

[the remark this answers is not in the document]
Comment (blink1073) [reply test/test_bson.py:987]: I don't think so.  If we had been going the other direction we could have specified `int()` formatting, but we'd have to basically recreate the old `str()` repr ourselves:

```python
f'{DatetimeConversion.DATETIME_AUTO}'
'4'
```

[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply test/test_bson.py:987]: I think we can use `class DatetimeConversion(int, enum.Enum)` to keep the old behavior.

File: pymongo/topology_description.py [hunk 2]
```diff
@@ -494,7 +494,7 @@ def _updated_topology_description_srv_polling(topology_description, seedlist):
         new_hosts = set(seedlist) - set(sds.keys())
         n_to_add = topology_description.srv_max_hosts - len(sds)
         if n_to_add > 0:
-            seedlist = sample(new_hosts, min(n_to_add, len(new_hosts)))
+            seedlist = sample(sorted(new_hosts), min(n_to_add, len(new_hosts)))
```
Comment (ShaneHarvey) [inline pymongo/topology_description.py:497]: Why use sorted() instead of list()?

**Review by ShaneHarvey — COMMENTED** [review]
One last question

File: pymongo/topology_description.py (hunk 2 above)
Comment (blink1073) [reply pymongo/topology_description.py:497]: It was recommended in the original traceback, I assume in case order was important.

**Comment by ShaneHarvey** [comment]
```
 [2022/10/11 16:54:43.381] FAIL [0.001s]: test_must_connect (test_client_context.TestClientContext.test_must_connect)
 [2022/10/11 16:54:43.381] ----------------------------------------------------------------------
 [2022/10/11 16:54:43.381] Traceback (most recent call last):
 [2022/10/11 16:54:43.381]   File "/data/mci/6d8fe5eaf124547d0611aa2dab642ffd/src/test/test_client_context.py", line 28, in test_must_connect
 [2022/10/11 16:54:43.381]     self.assertTrue(
 [2022/10/11 16:54:43.381] AssertionError: False is not true : client context must be connected when PYMONGO_MUST_CONNECT is set. Failed attempts:
 [2022/10/11 16:54:43.381] failed to connect client MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True, serverselectiontimeoutms=5000): 'bytes' object has no attribute 'is_absolute', Timeout: 5.0s, Topology Description: <TopologyDescription id: 63459f235735d3dee7ba4219, topology_type: Unknown, servers: [<ServerDescription ('localhost', 27017) server_type: Unknown, rtt: None, error=AttributeError("'bytes' object has no attribute 'is_absolute'")>]>
 [2022/10/11 16:54:43.381] failed to connect client MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True, serverselectiontimeoutms=5000, tls=True, tlscertificatekeyfile='/data/mci/6d8fe5eaf124547d0611aa2dab642ffd/src/test/certificates/client.pem', tlscafile='/data/mci/6d8fe5eaf124547d0611aa2dab642ffd/src/test/certificates/ca.pem'): 'bytes' object has no attribute 'is_absolute', Timeout: 5.0s, Topology Description: <TopologyDescription id: 63459f285735d3dee7ba421a, topology_type: Unknown, servers: [<ServerDescription ('localhost', 27017) server_type: Unknown, rtt: None, error=AttributeError("'bytes' object has no attribute 'is_absolute'")>]>
```

This error was fixed by PYTHON-3451.

File: pymongo/topology_description.py (hunk 2 above)
Comment (ShaneHarvey) [reply pymongo/topology_description.py:497]: Okay I don't think the sort is important but it doesn't hurt either besides a few wasted cycles.

File: .evergreen/run-tests.sh [hunk 3]
```diff
@@ -220,5 +220,9 @@ if [ -z "$GREEN_FRAMEWORK" ]; then
     $PYTHON $COVERAGE_ARGS setup.py $C_EXTENSIONS test $TEST_ARGS $OUTPUT
 else
     # --no_ext has to come before "test" so there is no way to toggle extensions here.
+    # Install the framework from pip.
+    # Use the same flags as the ones used by gevent, in case we have to build the wheel.
+    $PYTHON -m pip install --upgrade pip wheel setuptools
+    $PYTHON -m pip install $GREEN_FRAMEWORK
```
Comment (ShaneHarvey) [inline .evergreen/run-tests.sh:226]: Is this using a venv? If not we need to create one first before pip installing anything.

**Review by ShaneHarvey — COMMENTED** [review]
Could we merge this PR and push out the 3.11 gevent testing fix and the 3.11 macos wheels to a new ticket?

**Comment by blink1073** [comment]
> Could we merge this PR and push out the 3.11 gevent testing fix and the 3.11 macos wheels to a new ticket?

Yep, sounds good

File: .evergreen/run-tests.sh (hunk 3 above)
Comment (blink1073) [reply .evergreen/run-tests.sh:226]: It was not, I'll add a note in the new ticket

**Comment by blink1073** [comment]
I opened https://jira.mongodb.org/browse/PYTHON-3471 and https://jira.mongodb.org/browse/PYTHON-3470

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**